junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TypeBasedParameterResolver.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.engine.extension;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+/**
+ * {@link ParameterResolver} adapter which resolve a parameter based on its exact type.
+ * @param <T> the type of the parameter to resolve
+ * @since 5.6
+ */
+public abstract class TypeBasedParameterResolver<T> implements ParameterResolver {
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		return getParameterType(parameterContext).equals(enclosedTypeOfParameterResolver());
+	}
+
```

Comment:
I'd override `resolveParameter()` as well and have subclasses implement s/t like `public T resolve()`.

Reply:
Good point, I'm working on it.

Reply:
done, looks like one of the builds timed out

Reply:
Restarted the check.

Reply:
Which didn't work out. Let's ignore this "failure" for the time being.